### PR TITLE
Store reindex request in index

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently the reindex request is propogated to the coordinating node in
a cluster state update. This commit moves the reindex request into a
specialized reindex index.

### CODE REVIEW

**Comment by tbrooks8** [comment]
@ywelsch - POC for this. I have a couple of concerns currently:

1. Security is an issue. Currently, I am just mashing it under the task origin. But I assume that we will need to add a reindex origin?

2. I am uncertain if there are concerns that the Index request will return on the master (this is still currently a transport master action), but the get request may not find the doc immediately on the coordinator?

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java [hunk 1]
```diff
@@ -141,6 +168,56 @@ protected ClusterBlockException checkBlock(StartReindexJobAction.Request request
...
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            });
+        } else {
+            CreateIndexRequest createIndexRequest = new CreateIndexRequest();
+            createIndexRequest.settings(reindexIndexSettings());
+            createIndexRequest.index(ReindexTask.REINDEX_INDEX);
+            createIndexRequest.cause("auto(reindex api)");
```
Comment (ywelsch) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java]: we should not have any analyzed fields on this index (at least for now), as we do not do any searches, just lookup by ID, so perhaps set `"dynamic": false` for the mapping.

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexBasicTests.java [hunk 2]
```diff
@@ -108,7 +108,8 @@ public void testCopyManyPersistent() throws Exception {
         copy.source().setSize(5);
         BulkByScrollResponse reindexResponse = client().execute(StartReindexJobAction.INSTANCE, request).get().getReindexResponse();
         assertThat(reindexResponse, matcher().created(max).batches(max, 5));
-        assertHitCount(client().prepareSearch("dest").setSize(0).get(), max);
+        // TODO: Added assertBusy
+        assertBusy(() -> assertHitCount(client().prepareSearch("dest").setSize(0).get(), max));
```
Comment (ywelsch) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexBasicTests.java]: what required this change?

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java [hunk 3]
```diff
@@ -141,6 +168,56 @@ protected ClusterBlockException checkBlock(StartReindexJobAction.Request request
         return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
     }
 
+    private void createReindexTaskDoc(String taskId, ReindexRequest reindexRequest, boolean indexExists, ActionListener<Void> listener) {
+        if (indexExists) {
+            IndexRequest indexRequest = new IndexRequest(ReindexTask.REINDEX_INDEX).id(taskId);
```
Comment (ywelsch) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java]: use OpType.CREATE as well?

**Review by ywelsch — COMMENTED** [review]
I've left some preliminary comments.
re 1: yes, I think we should have a new REINDEX origin
re 2: that should not happen. If index is acknowledged, get should return the document on the coordinating node. The only perhaps a bit unfortunate thing is that the index will internally be refreshed to access the document. As the number of reindex requests to a cluster shouldn't be too high, I don't think this will be a problem.

**Comment by tbrooks8** [comment]
@elastic/es-security - Can someone from the security team take a look at this PR and give us some direct on how this should work with security? We need to be able to read and write from a specialized "reindex" index. I added a reindex origin similar to how we do with tasks, rollups, and data frames. Let me know if this is the correct approach.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java [hunk 4]
```diff
@@ -53,7 +69,9 @@ public TransportStartReindexJobAction(TransportService transportService, ThreadP
         super(StartReindexJobAction.NAME, transportService, clusterService, threadPool, actionFilters, StartReindexJobAction.Request::new,
             indexNameExpressionResolver);
         this.persistentTasksService = persistentTasksService;
-        this.client = client;
+
+        // TODO: Need reindex origin probably.
```
Comment (tvernum) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java]: You can drop this todo now :)

**Comment by tvernum** [comment]
That origin approach seems right to me.
With the number of origin's we're accumulating, it would be good to start running them with different privileges, but that's a separate issue.

**Comment by tbrooks8** [comment]
thanks @tvernum

**Review by ywelsch — APPROVED** [review]
Looks good. There is a relevant CI failure on the PR though. Also, can you add a TODO to the meta issue that we should look at making the `.reindex` index an alias instead that points to a `.reindex-7` index, similar as we do for the security index and other indices. This will allow us to more easily do upgrades in the future

**Pull request merged**